fix: Allocate tax loss to tax account head on early payment discount

marination opened the pull request:
Docs update: 
https://docs.erpnext.com/docs/v14/user/manual/en/accounts/payment-terms/edit-wiki?wiki_page_patch=9f963e285a
https://docs.erpnext.com/docs/v14/user/manual/en/accounts/accounts-settings/edit-wiki?wiki_page_patch=7b3e920d00

## Early Payment Discount: Loss Allocation
Early Payment Discount Deductions in payment entry can be split into income loss and tax loss by enabling **Book Tax Loss on Early Payment Discount** in the Accounts Settings:
  <img width="1313" alt="Tax_Settings_Revised" src="https://user-images.githubusercontent.com/25857446/228448996-427c1b84-e404-4d69-b1c4-37ed89973949.png">


**Consider the foll.g Sales Invoice:**
<img width="1297" alt="Screenshot 2023-03-03 at 1 39 58 PM" src="https://user-images.githubusercontent.com/25857446/222666573-5bd24e61-9f2f-400f-9ba8-d5f2de1bb718.png">


**Payment Entry Before:**
<img width="1048" alt="Screenshot 2023-03-03 at 1 34 54 PM" src="https://user-images.githubusercontent.com/25857446/222665428-a601bb91-c245-4cdd-8497-f1ec3b4b766d.png">


**Payment Entry After:**
<img width="1029" alt="deductions" src="https://user-images.githubusercontent.com/25857446/222665912-e7b1fbde-a288-4503-a128-7cc07f13c060.png">

## Other Fixes

1. **Deductions amount value must always be in terms of base currency/company currency**: Calculate and convert the deductions amount into base currency. They were being set in the Supplier/Customer currency which is wrong since these deductions are on the Company account heads
2. **Back update `discounted amount` in Invoice's Payment Schedule rows based on Discount Type**: Discounted Amount was always calculated in percentages on submitting a Payment Entry which gave the wrong value if it was an amount based discount.
3. **Set label and value formatting currency of `Amount` in deductions as base currency/company currency**: For multi-currency SI and PE, the deductions table's amount field label showed the company currency but the value was formatted in the Customer/Supplier currency
4. **If `Allocate Payment Based on Payment Terms` is checked, then handle payment schedule amount extraction and back updation in the right currency**: 
     - Payment Schedule Amounts are always in transaction currency. 
     - While fetching amount from Payment Schedule in the PE, if accounting is in base currency, convert the value into base currency.
     - Similarly while back **Paid Amount** into Payment Schedule, convert back into trans.n currency IF reference doc is multi-currency and PE is not (PE is in base currency)
5. **Provision to apply discount on recording payment late**: Party could have paid on time but payment is recorded late. Prompt for reference date so that discount is applied while mapping. **This is only prompted if early payment discount is applied on current document.**
    ![2023-03-14 13 49 25](https://user-images.githubusercontent.com/25857446/224939011-a5fc751e-9d52-4904-ad37-fe7aa6079d8c.gif)

<hr>

**TODO:**

- [x] Check with multi currency
- [x] Tests for discount %, discount amount and multi currency
- [x] Docs

marination committed:
fix: Taxes aren't discounted on early payment discount

- Deductions in payment entry must be split into income loss and tax loss
- Compute total discount in percentage, makes discounting different amounts proportionately easier

marination committed:
Merge branch 'develop' into early-payment-loss

marination committed:
fix: Recalculate difference amount after setting deductions

marination committed:
fix: Set deductions in base currency

- Use field precision to get more accurate values

marination committed:
fix: Back update discounted amount in Invoice based on discount type

- Discount value was always trated as a percentage on back updation

marination committed:
test: PE from SI with early payment discount amount & PE assertions in discount % test

marination committed:
Merge branch 'develop' into early-payment-loss

marination committed:
fix: Set deduction amount in company currency on Doctype

- Even via JS, deductions amount is always in company currency
- Since there is nothing dynamic about this field, set it in the doctype spec itself
- fixed: Inconsistency between label currency and field currency formatted value

barredterra committed:
Merge branch 'develop' into early-payment-loss

barredterra commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -1911,13 +1930,99 @@ def apply_early_payment_discount(paid_amount, received_amount, doc):
+
+
+def set_early_payment_discount_loss(pe, doc, valid_discounts, discount_amount):
+	"""Split early bird discount deductions into Income Loss & Tax Loss."""
```
```suggestion
	"""Split early payment discount into Income Loss & Tax Loss."""
```

barredterra commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -1911,13 +1930,99 @@ def apply_early_payment_discount(paid_amount, received_amount, doc):
+			tax_discount_loss[account] += tax_loss
+
+	for account, loss in tax_discount_loss.items():
+		total_tax_loss += loss
+		pe.append(
+			"deductions",
+			{
+				"account": account,
+				"cost_center": pe.cost_center or frappe.get_cached_value("Company", pe.company, "cost_center"),
+				"amount": flt(loss * doc.get("conversion_rate", 1), precision),
+			},
+		)
```
```suggestion
		total_tax_loss += loss
		amount = flt(loss * doc.get("conversion_rate", 1), precision)
		if amount == 0.0:
			continue

		pe.append(
			"deductions",
			{
				"account": account,
				"cost_center": pe.cost_center or frappe.get_cached_value("Company", pe.company, "cost_center"),
				"amount": amount,
			},
		)
```

Don't add deduction if the `amount == 0`. This can happen when the Sales Taxes and Charges table contains rows that don't apply to any item on that invoice.

barredterra reviewed (commented):
I tried to create PE from a SI in company currency

- ... with multiple tax rows and early payment discount
- ... with a single tax row and no early payment discount

This worked very well 👍🏼 

What didn't work correctly was to create a SI in another currency, while the GL entries use base amounts / company currency. In this case, the paid amount (in company currency) was wrong.
<details>
<summary>SI</summary>

![Bildschirmfoto 2023-03-06 um 22 56 27](https://user-images.githubusercontent.com/14891507/223244170-5ad83786-1e5c-4c72-bc3f-e9b2dfa34ca9.png)

</details>
<details>
<summary>PE</summary>

_Paid Amount (EUR)_ is wrong, deductions are correct

![Bildschirmfoto 2023-03-06 um 22 57 02](https://user-images.githubusercontent.com/14891507/223244174-a801317a-09ae-4e92-be24-73e377bd8011.png)

![Bildschirmfoto 2023-03-06 um 22 57 20](https://user-images.githubusercontent.com/14891507/223244176-db997426-d966-417d-bcdb-b7686a7bb9cc.png)

</details>

marination committed:
fix: Don't add to deductions if amount is 0

- misc: better docstring

marination committed:
fix: Paid amount must be discounted considering accounting currency

- Accounting is in the same currency if party currency and company currency is the same
- If accounting is in the same currency, paid and recvd amount is in the base currency
- Then, discount amount must also be in the base currency as it is deducted from paid amount
- Received amount must be in base currency if not multi currency
- cleanup: Deductions setting broken into smaller functions

marination committed:
fix: Multi-currency SI with base currency PE

- Return total discount loss in base currency
- Allocate payment based on terms: Set allocated amount in references table in base currency if accounting is in that currency
- Allocate payment based on terms: While back updating set paid amount (payment schedule) in transaction currency always
- minor: discount msgprint in correct currency

marination committed:
test: Multi currency SI with multi-currency accounting and single currency accounting + Early payment discount

marination committed:
Merge branch 'develop' into early-payment-loss

barredterra commented on erpnext/accounts/doctype/payment_entry/payment_entry.py:
```diff
@@ -1889,20 +1931,26 @@ def set_paid_amount_and_received_amount(
 
 	if doc.doctype in eligible_for_payments and has_payment_schedule:
 		for term in doc.payment_schedule:
 			if not term.discounted_amount and term.discount and getdate(nowdate()) <= term.discount_date:
```
```suggestion
			if not term.discounted_amount and term.discount and doc.reference_date <= term.discount_date:
```

The relevant date for determining if the discount applies is the date the transaction was received. If this would make things easier, we could argue that the posting date would work as well – I guess they can be identical for most practical purposes.

Imagine the customer paid yesterday, but I'm recording the payment today. The discount should still apply, if yesterday was the last day of validity.

Another possibility would be to ask the user whether the early payment discount should be applied. This would help in cases where the company accepts discounted payments even after the validity date, to avoid trouble.

marination replied:
I think reference date would be safest since it's purpose is to specify the actual cheque date or receipt date but it is not set until after the mapping occurs :(

barredterra reviewed (commented):
Issues from the last review have been resolved. 👍🏼 

Something that should go into the docs is that this feature (currently) only works when mapping a payment from an invoice, not when creating a payment and then fetching outstanding invoices.

One small annoyance I encountered was that with odd amounts I can get an additional deduction of `0.01`. For example, when two deductions have been rounded down but the difference adds up to > `0,005`. This deduction of `0.01` has no account set. We could either set the company's _Default Round Off Account_ here or change the calculation method to avoid rounding differences. This can be recreated with the example SI and PE from the last review.

marination committed:
fix: Handle rounding more gracefully

- Round off pending discount loss to avoid miniscule losses rounded to 0.0 that are added in deductions
- Use base amounts to calculate base losses instead of using conversion factor which increases rounding error
- Round of total base loss instead of individual income and tax losses to reduce rounding error
- Use default round off account for pending rounding loss in deductions

marination committed:
Merge branch 'develop' into early-payment-loss

marination committed:
fix: Provision to apply early payment discount if payment is recorded late

- Party could have paid on time but payment is recorded late
- Prompt for reference date so that discount is applied while mapping
- Prompt only if discount in payment schedule of valid doctypes
- test: Reference date and impact on PE
- `make_payment_entry` (JS) must be able to access `this`

marination committed:
Merge branch 'develop' into early-payment-loss

marination commented:
@deepeshgarg007 Mind taking a look 👀?

deepeshgarg007 reviewed (commented):
The tax loss booking should be optional IMO. In most case this is [Non Trade (Cash Discount)](https://byjus.com/commerce/difference-between-trade-discount-and-cash-discount/) and don't affect tax, at least this how it is in India

Also the docs link says v13, should be v14

barredterra commented:
A supplier can reduce their GST liability when granting a cash discount, provided certain conditions are met. According to Section 15(3) of the CGST Act, 2017, the value of the supply shall not include any discount if:

1. The discount is given before or at the time of supply and has been duly recorded in the invoice issued in respect of such supply.
2. The discount is given after the supply has been effected, if:
    a) The discount is established in terms of an agreement entered into at or before the time of such supply and specifically linked to relevant invoices.
    b) The input tax credit attributable to the discount, on the basis of the document issued by the supplier, has been reversed by the recipient of the supply.

By meeting these conditions, the supplier can effectively reduce the taxable value of the supply, resulting in a lower GST liability. However, it is essential to ensure that all applicable conditions are met and necessary documentation is maintained to support the reduction in GST liability.

Summarized from https://carajput.com/learn/whether-liable-to-gst-on-discount-received-for-early-payments.html

---

Under the GST regime, Section 15 of the CGST Act, 2017 governs discounts. Discounts given before or at the time of supply can be reduced from the taxable value if mentioned on the invoice. Discounts given after the supply can also be reduced from the taxable value if the discount is agreed upon before the supply, linked to a specific supply invoice, and the Input Tax Credit (ITC) attributable to the discount is reversed by the buyer. GST liability of the supplier can be reduced if both the supplier and receiver are aware of the discount before supply. There is no differentiation between trade and cash discounts in GST; instead, they are segregated into two categories: those given before or at the time of supply, and those given after the time of supply.

Summarized from https://taxguru.in/goods-and-service-tax/tax-treatment-tradecash-discounts-gst-regime.html

deepeshgarg007 commented:
@barredterra Didn't exactly get your last response, are you suggesting the tax loss booking should not be optional?

barredterra commented:
I couldn't access the information you pointed to (blocked in Europe), but the articles I found seemed to support our use case. ("GST liability of the supplier can be reduced if both the supplier and receiver are aware of the discount before supply. There is no differentiation between trade and cash discounts in GST")

Anyway, making it configurable shouldn't be an obstacle. First rule of ERPs: everything must be configurable. 😅

marination committed:
feat: Make Tax loss booking optional

- Checkbox in Accounts Settings
- Apply checkbox in PE deductions setting logic
- Adjust tests

marination committed:
Merge branch 'develop' into early-payment-loss

marination commented:
@deepeshgarg007 It is configurable now in the Accounts Settings (see description). Do let me know if any other major changes are needed
